### PR TITLE
Allow changing DebugLog severity during runtime + some tweaks

### PR DESCRIPTION
**Pull request by olanti-p**
#### Purpose of change
As @Coolthulhu mentioned in https://github.com/cataclysmbnteam/Cataclysm-BN/pull/525#issuecomment-842450063 , our logging system - DebugLog - in extremely awkward to use.

Debug levels and classes are enabled via preprocessor macros, so there's no way to change them once the game is compiled.
This could've been explained by performance concerns, as in theory, compiler should optimize away disabled log statements.
But after some testing and profiling it turned that neither Visual Studio with Release configuration nor `gcc` with `RELEASE=1 LTO=1` are smart enough to do that.

As such, there should be no harm in unhardcoding debug severity.

#### Describe the solution
1. Change debug levels & classes to be proper enums, and use `enum_bitset` to track currently enabled levels & classes. This makes it no longer possible to define message as belonging to multiple levels/classes, but such functionality was used only once anyway.
2. Remove `#define`s from the Makefile, and instead add a bunch of boolean options for controlling each class & level.
3. Add new debug class for exclusively logging `debugmsg`s.
4. Add new debug class for logging messages that show up in the sidebar when debug mode is ON.
5. Rename "PEDANTIC_INFO" debug level to "Debug".
6. Change all usages of "DC_ALL" and "DL_ALL" into something more appropriate.
7. Tweak some `DebugLog`/`debugmsg` invocations to better fit their intended purpose.
8. Make `DebugLog` append newline and flush at the end of each line rather than at the beginning. This allows interleaving `DebugLog` and `cata_printf` statements in test mode without turning the result into a mess, and seeing messages immediately when using the log for live debugging.
9. Allow `DebugLog` to capture file name & line number (as a new `DebugLogFL` macro).
10. Implement collapsible groups for option manager, and use such group for DebugLog settings.

~~All actual changes are in the last 3 commits, the rest - from #554. #554 is not required for this one, I've just used it to avoid merge conflicts.~~ Force-pushed.

#### Describe alternatives you've considered
1. Switch to some lightweight logging library? Fairly sure there are tons of great ones out there.
2. More debug classes?
3. Separate debug class for JSON? Should make it possible to report JSON problems with varying severity
4. Allow disabling Error level? I'm somewhat concerned about people mistakenly disabling it and then submitting useless logs...

#### Testing
The log looks more or less how it used to.
Tileset loading report is properly flushed (it used to require additional empty line for that).
When running tests, messages from DebugLog look fine when interleaved with regular output:
<details>

  <summary> Example output from local run </summary>

  ```
 . . .
0.000 s: address should be same when translation is absent
0.000 s: translations_macro_char_address
0.000 s: current address of the arg should be returned when translation is absent
0.000 s: translations_macro_char_address
15:34:42.254 INFO : Language is set to: 'en_US'/'en_US'
15:34:42.255 INFO : C locale set to 'en_US.UTF-8'
15:34:42.255 INFO : C++ locale set to 'en_US.UTF-8'
15:34:42.750 INFO : Language is set to: 'fr_FR'/'fr_FR'
15:34:42.750 INFO : C locale set to 'fr_FR.UTF-8'
15:34:42.750 INFO : C++ locale set to 'fr_FR.UTF-8'
15:34:43.230 INFO : Language is set to: 'ru_RU'/'ru_RU'
15:34:43.230 INFO : C locale set to 'ru_RU.UTF-8'
15:34:43.231 INFO : C++ locale set to 'ru_RU.UTF-8'
15:34:43.581 INFO : Using experimental system, language set to 'en_US'
15:34:43.581 INFO : C locale set to 'en_US.UTF-8'
15:34:43.581 INFO : C++ locale set to 'en_US.UTF-8'
15:34:44.137 INFO : Using experimental system, language set to 'fr_FR'
15:34:44.137 INFO : C locale set to 'fr_FR.UTF-8'
15:34:44.137 INFO : C++ locale set to 'fr_FR.UTF-8'
15:34:44.628 INFO : Using experimental system, language set to 'ru_RU'
15:34:44.628 INFO : C locale set to 'ru_RU.UTF-8'
15:34:44.628 INFO : C++ locale set to 'ru_RU.UTF-8'
15:34:45.023 INFO : Language is set to: 'en_US'/'en_US'
15:34:45.023 INFO : C locale set to 'en_US.utf-8'
15:34:45.023 INFO : C++ locale set to 'en_US.utf-8'
3.250 s: translations_actually_translate
0.000 s: units_have_correct_ratios
0.000 s: energy parsing from JSON
0.000 s: time_duration parsing from JSON
0.000 s: value_ptr copy constructor
 . . .
  ```

</details>

Not sure what's up with Travis, as it appears to be always buffering `stdout` but immediately printing `stderr`:
<details>

  <summary> Example output from Travis </summary>

  ```
 . . .
12:37:52.010 WARNING : opendir [modded/gfx/] failed with "No such file or directory".
12:37:52.029 WARNING : opendir [modded/mods/] failed with "No such file or directory".
12:37:53.862 WARNING : opendir [modded/save/Tom/mods/] failed with "No such file or directory".
// Tests started
// stderr output from "translations_actually_translate" test
12:38:48.099 INFO : Language is set to: 'en_US'/'en_US'
12:38:48.099 INFO : C locale set to 'en_US.UTF-8'
12:38:48.099 INFO : C++ locale set to 'en_US.UTF-8'
12:38:48.311 INFO : Language is set to: 'fr_FR'/'fr_FR'
12:38:48.312 INFO : C locale set to 'fr_FR.UTF-8'
12:38:48.312 INFO : C++ locale set to 'fr_FR.UTF-8'
12:38:48.522 INFO : Language is set to: 'ru_RU'/'ru_RU'
12:38:48.525 INFO : C locale set to 'ru_RU.UTF-8'
12:38:48.525 INFO : C++ locale set to 'ru_RU.UTF-8'
12:38:48.669 INFO : Using experimental system, language set to 'en_US'
12:38:48.669 INFO : C locale set to 'en_US.UTF-8'
12:38:48.669 INFO : C++ locale set to 'en_US.UTF-8'
12:38:48.901 INFO : Using experimental system, language set to 'fr_FR'
12:38:48.901 INFO : C locale set to 'fr_FR.UTF-8'
12:38:48.901 INFO : C++ locale set to 'fr_FR.UTF-8'
12:38:49.116 INFO : Using experimental system, language set to 'ru_RU'
12:38:49.116 INFO : C locale set to 'ru_RU.UTF-8'
12:38:49.116 INFO : C++ locale set to 'ru_RU.UTF-8'
12:38:49.283 INFO : Language is set to: 'en_US'/'en_US'
12:38:49.283 INFO : C locale set to 'en_US.UTF-8'
12:38:49.283 INFO : C++ locale set to 'en_US.UTF-8'

// stdout output from tests - buffered and printed at the end of the run?
Starting the actual test at Tue Jun  1 12:37:59 2021
0.169 s: place_active_item_at_various_coordinates
0.141 s: load_all_base_game_mos
0.261 s: recipe_permutations
0.322 s: soldier zombie
0.322 s: effective vs actual damage per second
0.555 s: smoker zombie
0.555 s: effective vs actual damage per second
0.507 s: survivor zombie
0.507 s: effective vs actual damage per second
0.323 s: soldier zombie
0.324 s: accuracy increases success
0.625 s: smoker zombie
0.626 s: accuracy increases success
0.442 s: survivor zombie
0.442 s: accuracy increases success
0.383 s: grid_and_vehicle_outside_bubble
0.503 s: grid_and_vehicle_outside_bubble
0.420 s: grid_and_vehicle_outside_bubble
0.350 s: grid_and_vehicle_outside_bubble
0.607 s: tripoint_hash_distribution
0.254 s: line_to_boundaries
0.246 s: map_bounds_checking
0.390 s: place_player_can_safely_move_multiple_submaps
0.487 s: monster_speed_square
0.580 s: monster_speed_trig
17.544 s: starting_items
0.578 s: npc_can_target_player
1.922 s: default_overmap_generation_always_succeeds
1.181 s: default_overmap_generation_has_non_mandatory_specials_at_origin
0.181 s: Exactly one endgame lab finale is generated in 0,0 overmap
0.187 s: Lightly clothed target temperatures
0.198 s: Player body temperatures within expected bounds.
0.398 s: Heavily clothed target temperatures
0.409 s: Player body temperatures within expected bounds.
0.418 s: Arctic gear target temperatures
0.429 s: Player body temperatures within expected bounds.
3.470 s: an unskilled shooter with an inaccurate shotgun
3.480 s: unskilled_shooter_accuracy
0.385 s: x_in_y_distribution
0.449 s:      When: the avatar runs the required distance
0.449 s:     Given: a new game
0.449 s: achievement_marathon
0.449 s: movement
0.449 s: achievments_tracker
0.198 s:      When: the avatar does the required actions
0.198 s:     Given: a new game
0.198 s: achievement_swim_merit_badge
0.198 s: movement
0.198 s: achievments_tracker
2.138 s: starve_test
0.907 s: starve_test_hunger3
5.669 s: all_nutrition_starve_test
0.290 s: string_ids_intern_test
0.113 s: test_player_vs_zombie_rock_basestats
0.125 s: basic_throwing_sanity_tests
0.159 s: test_player_vs_zombie_javelin_iron_basestats
0.170 s: basic_throwing_sanity_tests
0.249 s: test_player_vs_zombie_rock_athlete
0.261 s: basic_throwing_sanity_tests
0.315 s: test_player_vs_zombie_javelin_iron_athlete
0.327 s: basic_throwing_sanity_tests
0.110 s: mid_skill_basestats_rock
0.123 s: throwing_skill_impact_test
0.536 s: test_player_kills_zombie_with_rock_basestats
0.547 s: player_kills_zombie_before_reach
1.392 s: translations_actually_translate                     // <- actual test that produced log messages
0.109 s: car on pavement
0.109 s: vehicle_efficiency
0.109 s: car on dirt
0.109 s: vehicle_efficiency
0.122 s: suv on pavement
 . . .
  ```

</details>

### CODE REVIEW

**Comment by olanti-p** [comment]
Ready.

Used the new option grouping functionaity to hide font settings in graphics tab, as it's low priority but takes a lot of space.

File: src/map.cpp [hunk 1]
```diff
@@ -6736,21 +6731,26 @@ void map::loadn( const tripoint &grid, const bool update_vehicles )
     static const oter_id rock( "empty_rock" );
     static const oter_id air( "open_air" );
 
-    dbg( D_INFO ) << "map::loadn(game[" << g.get() << "], worldx[" << abs_sub.x
-                  << "], worldy[" << abs_sub.y << "], grid " << grid << ")";
+    /*
+    dbg( DL::Debug ) << "map::loadn( game=" << g.get()
+                     << ", world=" << abs_sub
+                     << ", grid=" << grid << " )";
+    */
```
Comment (Coolthulhu) [inline src/map.cpp:6738]: Why commented out?

File: src/map.cpp [hunk 2]
```diff
@@ -6736,21 +6731,26 @@ void map::loadn( const tripoint &grid, const bool update_vehicles )
...
 
     const tripoint grid_abs_sub = abs_sub.xy() + grid;
     const size_t gridn = get_nonant( grid );
 
-    dbg( D_INFO ) << "map::loadn grid_abs_sub: " << grid_abs_sub << "  gridn: " << gridn;
+    /*
+    dbg( DL::Debug ) << "map::loadn grid_abs_sub: " << grid_abs_sub << "  gridn: " << gridn;
+    */
```
Comment (Coolthulhu) [inline src/map.cpp:6745]: And here

File: src/mapbuffer.cpp [hunk 3]
```diff
@@ -89,14 +87,14 @@ void mapbuffer::remove_submap( tripoint addr )
 
 submap *mapbuffer::lookup_submap( const tripoint &p )
 {
-    dbg( D_INFO ) << "mapbuffer::lookup_submap( x[" << p.x << "], y[" << p.y << "], z[" << p.z << "])";
+    //DebugLog( DL::Debug, DC::Map ) << "mapbuffer::lookup_submap " << p;
```
Comment (Coolthulhu) [inline src/mapbuffer.cpp:90]: And here

**Comment by Coolthulhu** [comment]
Overall looks really good, like a chance to reclaim the debug log system into something that could help find the bugs.

**Comment by olanti-p** [comment]
Commented out entries are so frequently called they show up on profiler when walking in a straight line.
Those 3 sum up to ≈1%. It's not much admittedly... Should I uncomment them?

**Comment by Coolthulhu** [comment]
I'd say remove. If they're too expensive, it's not a good idea to keep them.
Though sometimes knowing that crash happens between certain two submaps is useful.

**Comment by olanti-p** [comment]
Done.

**Pull request merged**